core/src/avm2/optimizer/optimize.rs

```diff
@@ -79,54 +79,48 @@ impl<'gc> OptValue<'gc> {
         matches!(self.null_state, NullState::IsNull)
     }
 
-    pub fn not_null(self, activation: &mut Activation<'_, 'gc>) -> bool {
+    pub fn not_null(self) -> bool {
         if matches!(self.null_state, NullState::NotNull) {
-            return true;
+            true
+        } else {
+            // Primitives are always not-null
+            self.class.is_some_and(|c| {
+                c.is_builtin_int()
+                    || c.is_builtin_uint()
+                    || c.is_builtin_number()
+                    || c.is_builtin_boolean()
+                    || c.is_builtin_void()
```

Comment:
This check appears often enough that it may be worth a dedicated `.is_primitive()` method.